src/dfx/src/commands/deps/pull.rs

```diff
@@ -0,0 +1,342 @@
+        let pulled_canister = pulled_json
+            .canisters
+            .get_mut(canister_id)
+            .ok_or_else(|| anyhow!("Failed to find {canister_id} entry in pulled.json"))?;
+        pulled_canister.name = Some(name.clone());
+    }
+
+    save_pulled_json(&project_root, &pulled_json)?;
+    Ok(())
+}
+
+#[context("Failed to fetch and parse `dfx:deps` metadata from canister {canister_id}.")]
+async fn fetch_deps_to_pull(
+    agent: &Agent,
+    logger: &Logger,
+    canister_id: Principal,
+    canisters_to_pull: &mut VecDeque<Principal>,
+    pulled_json: &mut PulledJson,
+) -> DfxResult {
+    info!(
+        logger,
+        "Resolving dependencies of canister {canister_id}..."
+    );
+
+    match fetch_metadata(agent, canister_id, DFX_DEPS).await? {
+        Some(deps_raw) => {
+            let deps_str = String::from_utf8(deps_raw)?;
+            let deps = parse_dfx_deps(&deps_str)?;
+            canisters_to_pull.extend(deps.iter().copied());
+            pulled_json.canisters.insert(
+                canister_id,
+                PulledCanister {
+                    deps,
+                    ..Default::default()
+                },
+            );
+        }
+        None => {
+            warn!(
+                logger,
+                "`{DFX_DEPS}` metadata not found in canister {canister_id}."
+            );
+            pulled_json
+                .canisters
+                .insert(canister_id, PulledCanister::default());
+        }
+    }
+    Ok(())
+}
+
+// download canister wasm, candid, init
+async fn download_canister_files(
+    env: &dyn Environment,
+    logger: &Logger,
+    canister_id: Principal,
+    pulled_canister: &mut PulledCanister,
+) -> DfxResult {
+    info!(logger, "Pulling canister {canister_id}...");
+
+    let agent = env
+        .get_agent()
+        .ok_or_else(|| anyhow!("Cannot get HTTP client from environment."))?;
+
+    // try fetch `dfx:wasm_hash`. If not available, get the hash of the on chain canister.
+    let hash_on_chain = match fetch_metadata(agent, canister_id, DFX_WASM_HASH).await? {
+        Some(wasm_hash_raw) => {
+            let wasm_hash_str = String::from_utf8(wasm_hash_raw)?;
+            trace!(
+                logger,
+                "Canister {canister_id} specified a custom hash: {wasm_hash_str}"
+            );
+            hex::decode(wasm_hash_str)?
+        }
+        None => {
+            match read_state_tree_canister_module_hash(agent, canister_id).await? {
+                Some(hash_on_chain) => hash_on_chain,
+                None => {
+                    bail!("Canister {canister_id} doesn't have module hash. Perhaps it's not installed.");
+                }
+            }
+        }
+    };
+
+    pulled_canister.wasm_hash = hex::encode(&hash_on_chain);
+
+    // will save wasm and candid in $(cache_root)/pulled/{canister_id}/
+    let canister_dir = get_cache_root()?
+        .join("pulled")
+        .join(canister_id.to_string());
```

Comment:
There are three methods (`get_pulled_wasm_path`, `get_pulled_service_candid_path`, and `get_pulled_wasm_url_txt_path`) that share this logic.  I'd extract this to a method and have all of them call it.